File: src/main/java/ukalus/container/ObjectRegistry.java
/*
 * Created on Jan 25, 2005
 *
 */
package ukalus.container;

import ukalus.container.transformers.BooleanTransformer;
import ukalus.container.transformers.ByteTransformer;
import ukalus.container.transformers.CharTransformer;
import ukalus.container.transformers.ClassTransformer;
import ukalus.container.transformers.DoubleTransformer;
import ukalus.container.transformers.FloatTransformer;
import ukalus.container.transformers.IntTransformer;
import ukalus.container.transformers.LongTransformer;
import ukalus.container.transformers.ShortTransformer;
import ukalus.container.transformers.StringTransformer;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * @author gremopm
 * 
 */
public class ObjectRegistry {

  private Map<Class, Function> converters = new HashMap<Class, Function>();
  private Map<String, ObjectFactory> factories = new HashMap<String, ObjectFactory>();

  {
    converters.put(Boolean.TYPE, new BooleanTransformer());
    converters.put(Byte.TYPE, new ByteTransformer());
    converters.put(Character.TYPE, new CharTransformer());
    converters.put(Class.class, new ClassTransformer());
    converters.put(Double.TYPE, new DoubleTransformer());
    converters.put(Float.TYPE, new FloatTransformer());
    converters.put(Integer.TYPE, new IntTransformer());
    converters.put(Long.TYPE, new LongTransformer());
    converters.put(Object.class, new StringTransformer());
    converters.put(Short.TYPE, new ShortTransformer());
    converters.put(String.class, new StringTransformer());
  }

  public void addConverter(Class type, Function converter) {
    converters.put(type, converter);
  }

  public Function getConverter(Class type) {
    return converters.get(type);
  }

  public void addObject(String id, ObjectFactory factory) {
    factories.put(id, factory);
  }

  public Object getObject(String id) throws IllegalArgumentException,
      InstantiationException, IllegalAccessException, InvocationTargetException {
    Object result = null;
    ObjectFactory factory = factories.get(id);
    if (factory != null) {
      result = factory.getInstance();
    }
    return result;
  }

}

File: src/main/kotlin/ukalus/container/ObjectRegistry.kt
/*
 * Created on Jan 25, 2005
 *
 */
package ukalus.container

import java.io.File

/**
 * @author gremopm
 */
class ObjectRegistry {

    private val converters = mutableMapOf<Class<*>, (Any) -> Any>(
            Boolean::class.javaPrimitiveType!! to { it -> (it as String).toBoolean() },
            Byte::class.javaPrimitiveType!! to { it -> (it as String).toByte() },
            Character::class.javaPrimitiveType!! to { it -> (it as String).toInt().toChar() },
            (Class::class.java as Class<*>) to { it -> Class.forName(it as String) },
            Double::class.javaPrimitiveType!! to { it -> (it as String).toDouble() },
            (File::class.java as Class<*>) to { it -> File(it as String) },
            Float::class.javaPrimitiveType!! to { it -> (it as String).toFloat() },
            Int::class.javaPrimitiveType!! to { it -> (it as String).toInt() },
            Long::class.javaPrimitiveType!! to { it -> (it as String).toLong() },
            (Any::class.java as Class<*>) to Any::toString,
            Short::class.javaPrimitiveType!! to { it -> (it as String).toShort() },
            (String::class.java as Class<*>) to { it -> it as String }
    )
    private val factories = mutableMapOf<String, ObjectFactory>()

    fun addConverter(type: Class<*>, converter: (Any) -> Any) {
        converters[type] = converter
    }

    fun getConverter(type: Class<*>): ((Any) -> Any)? = converters[type]

    fun addObject(id: String, factory: ObjectFactory) {
        factories[id] = factory
    }

    fun getObject(id: String): Any? = factories[id]?.instance

}